state-chain/pallets/cf-asset-balances/src/lib.rs

```diff
@@ -128,6 +146,21 @@ pub mod pallet {
 	#[pallet::storage]
 	pub type WithheldAssets<T: Config> =
 		StorageMap<_, Twox64Concat, Asset, AssetAmount, ValueQuery>;
+
+	#[pallet::storage]
+	/// Storage for user's free balances/ DoubleMap: (AccountId, Asset) => Balance
+	pub type FreeBalances<T: Config> =
+		StorageDoubleMap<_, Twox64Concat, T::AccountId, Identity, Asset, AssetAmount>;
+
+	#[pallet::storage]
+	/// Historical earned fees for an account.
+	pub type HistoricalEarnedFees<T: Config> =
+		StorageDoubleMap<_, Identity, T::AccountId, Twox64Concat, Asset, AssetAmount, ValueQuery>;
+
+	/// Fund accrued from rejected swap and CCM calls.
+	#[pallet::storage]
+	pub type CollectedRejectedFunds<T: Config> =
+		StorageMap<_, Twox64Concat, Asset, AssetAmount, ValueQuery>;
```

Comment:
I can see how it might make sense to track these funds here, but then we could track them in the WithheldAssets storage, for example. Not sure there's a real need for a distinct storage item.

Reply:
I think we have to keep these two things separate, or we have to book keep that we don't refund these funds to validators. That storage item only tracks failed CCM swaps. We also emit an event in this case with all information about failed swap, I guess we do this, so the emitter of the CCM swap has some kind of receipt he can use for a later refund.